Modify function-call cleanup

ichizok opened the pull request:
Ref: #3835

## Proposal

At the end of function call, if "funccall" object and its "l:" vars are not referred by anyone,
Vim may be able to free "a:", "a:000" vars, so and "funccall".

Therefore, this patch changes:

* In cleanup_function_call(), check the references of "l:", "a:", "a:000" separately and free each of them if can (or retain)

By doing this, in the case of #3835 "funccall" object can be freed just after its function ends,
so the memory usage doesn't gain almost and then needs no GC here.

## Confirmation

On Ubuntu 18.04.

test.vim
```vim
func s:f(...)
  let l:x = a:000
endfunc
for _ in range(100000)
  call s:f(0)
endfor
```

```
$ vim --clean test.vim
:so %
```

monitoring memory usage:

```
$ while :; do grep '^VmRSS' /proc/$(pgrep vim)/status; sleep 1; done
```

8.1.0918:
```
VmRSS:      8712 kB
VmRSS:      8712 kB
VmRSS:      8712 kB
VmRSS:    225716 kB    # do ":so %"
VmRSS:    225716 kB
VmRSS:    225716 kB
VmRSS:    225716 kB    # though GC was done, RSS doesn't decrease immediately because of cached
...
```

patched:
```
VmRSS:      8740 kB
VmRSS:      8740 kB
VmRSS:      8740 kB
VmRSS:     13224 kB    # do ":so %"
VmRSS:     13224 kB
VmRSS:     13224 kB
VmRSS:     13224 kB    # no change also after this
...
```

ichizok committed:
Fix: free used storages of funccall

when function-call finishes

ichizok committed:
Modify memory-allocation in call_user_func()

ichizok committed:
Modify "made_copy" limit: base on pagesize * 1024

ichizok committed:
Add test

ichizok committed:
Skip memory usage tests on ASAN build

brammool commented:
> ## Proposal
> 
> At the end of function call, if "funccall" object and its "l:" vars
> are not referred by anyone, Vim may be able to free "a:", "a:000"
> vars, so and "funccall".
> 
> Therefore, this patch changes:
> 
> * In cleanup_function_call(), check the references of "l:", "a:",
> "a:000" separately and free each of them if can (or retain)
> 
> By doing this, in the case of #3835 "funccall" object can be freed
> just after its function ends, so the memory usage doesn't gain almost
> and then needs no GC here.
> 
> ## Confirmation
> 
> On Ubuntu 18.04.
> 
> test.vim
> ```vim
> func s:f(...)
>   let l:x = a:000
> endfunc
> for _ in range(100000)
>   call s:f(0)
> endfor
> ```
> 
> ```
> $ vim --clean test.vim
> :so %
> ```
> 
> monitoring memory usage:
> 
> ```
> $ while :; do grep '^VmRSS' /proc/$(pgrep vim)/status; sleep 1; done
> ```
> 
> 8.1.0918:
> ```
> VmRSS:      8712 kB
> VmRSS:      8712 kB
> VmRSS:      8712 kB
> VmRSS:    225716 kB    # do ":so %"
> VmRSS:    225716 kB
> VmRSS:    225716 kB
> VmRSS:    225716 kB    # though GC was done, RSS doesn't decrease immediately because of cached
> ...
> ```
> 
> patched:
> ```
> VmRSS:      8740 kB
> VmRSS:      8740 kB
> VmRSS:      8740 kB
> VmRSS:     13224 kB    # do ":so %"
> VmRSS:     13224 kB
> VmRSS:     13224 kB
> VmRSS:     13224 kB    # no change also after this
> ...
> ```

Thanks for looking into this.  A few remarks:

The change to .travis.yml is needed?  Or was this just for testing?

I don't like renaming vars_clear_ext() to vars_clear_ht().
vars_clear also takes a hashtab_T argument, the difference is that the
_ext function also takes a flag, thus it's "extended".

-- 
Friends?  I have lots of friends!  In fact, I have all episodes ever made.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

ichizok commented:
> The change to .travis.yml is needed? Or was this just for testing?

Though this is not related to the subject of this pull-req, I felt it is too small to make a new pull-req for only this change so I included to this patch.

This change means:
Bash executes "while read" inside subprocess so lose variables at the end of loop, thus "log" variable is always empty outside the loop and this block exits with 0.
By finishing the inside of "do ~ done" by "false" it makes whole this block exit with 1.

> I don't like renaming vars_clear_ext() to vars_clear_ht().
> vars_clear also takes a hashtab_T argument, the difference is that the
> _ext function also takes a flag, thus it's "extended".

I understand. I updated the patch and reverted renaming.

ichizok commented:
I updated the patch: I removed the change for .travis.yml since because the additional changes (and I will create new pull-req for it).
